src/repr/src/relation.rs

```diff
@@ -657,13 +819,285 @@ impl fmt::Display for NotNullViolation {
+        let latest_version = self.latest_version();
+        let new_version = latest_version.bump();
+
+        let col = self
+            .inner
+            .metadata
+            .values_mut()
+            .find(|meta| meta.name == name)
+            .expect("column to exist");
+
+        assert_none!(col.dropped, "column was already dropped");
+        col.dropped = Some(new_version);
+
+        self.validate();
+        new_version
+    }
+
+    /// Returns the [`RelationDesc`] at the latest version.
+    pub fn latest(&self) -> RelationDesc {
+        self.inner.clone()
+    }
+
+    /// Returns this [`RelationDesc`] at the specified version.
+    pub fn at_version(&self, version: RelationVersionSelector) -> RelationDesc {
+        // Get all of the changes from the start, up to whatever version was requested.
+        //
+        // TODO(parkmycar): We should probably panic on unknown verisons?
+        let up_to_version = match version {
+            RelationVersionSelector::Latest => RelationVersion(u64::MAX),
+            RelationVersionSelector::Specific(v) => v,
+        };
+
+        let valid_columns = self.inner.metadata.iter().filter(|(_col_idx, meta)| {
+            let added = meta.added <= up_to_version;
+            let dropped = meta
+                .dropped
+                .map(|dropped_at| up_to_version >= dropped_at)
+                .unwrap_or(false);
+
+            added && !dropped
+        });
+
+        let mut column_types = Vec::new();
+        let mut column_metas = BTreeMap::new();
+
+        // N.B. At this point we need to be careful because col_idx might not
+        // equal typ_idx.
+        for (col_idx, meta) in valid_columns {
+            let new_meta = ColumnMetadata {
+                name: meta.name.clone(),
+                typ_idx: column_types.len(),
+                added: meta.added.clone(),
+                dropped: meta.dropped.clone(),
+            };
+            column_types.push(self.inner.typ.columns()[meta.typ_idx].clone());
+            column_metas.insert(*col_idx, new_meta);
+        }
+
+        let relation_type = RelationType {
+            column_types,
+            keys: self.inner.typ.keys.clone(),
+        };
+
+        RelationDesc {
+            typ: relation_type,
+            metadata: column_metas,
+        }
+    }
+
+    pub fn latest_version(&self) -> RelationVersion {
+        self.inner
+            .metadata
+            .values()
+            // N.B. Dropped is always greater than added.
+            .map(|meta| meta.dropped.unwrap_or(meta.added))
+            .max()
+            // If there aren't any columns we're implicitly the root version.
+            .unwrap_or(RelationVersion::root())
+    }
+
+    /// Validates internal contraints of the [`RelationDesc`] are correct.
+    ///
+    /// # Panics
+    ///
+    /// Panics if a constraint is not satisfied.
+    fn validate(&self) {
+        fn validate_inner(desc: &RelationDesc) -> Result<(), anyhow::Error> {
```

Comment:
Are there any validations to add on `desc.typ.keys`? Maybe that each index is valid and that a single key doesn't have duplicate indexes.

Reply:
Good call, added some validation